File: ios/Packages/AppFeatures/Sources/GamificationFeature/GamificationViewModel.swift
import CoreKit
import Foundation
import NetworkingKit
import Observation

/// State machine for the Gamification screen (docs/features/gamification.md).
/// Renders server descriptors verbatim — no client-side scoring/streak math.
@MainActor
@Observable
public final class GamificationViewModel {
    public private(set) var profile: GamificationProfile = .empty
    public private(set) var isLoading = false
    public private(set) var error: String?

    private let client: AuthenticatedAPIClientProtocol
    private let recorder: GamificationEventRecorder?
    private let timezone: String
    private let widgetStore: GamificationWidgetSnapshotStore?
    private let reloadWidgets: (@Sendable () -> Void)?
    private let now: @Sendable () -> Date

    public init(
        client: AuthenticatedAPIClientProtocol,
        recorder: GamificationEventRecorder? = nil,
        timezone: String = TimeZone.current.identifier,
        widgetStore: GamificationWidgetSnapshotStore? = nil,
        reloadWidgets: (@Sendable () -> Void)? = nil,
        now: @escaping @Sendable () -> Date = { Date() }
    ) {
        self.client = client
        self.recorder = recorder
        self.timezone = timezone
        self.widgetStore = widgetStore
        self.reloadWidgets = reloadWidgets
        self.now = now
    }

    public func load() async {
        isLoading = true
        error = nil
        // Flush any queued events first so a fresh profile reflects them.
        await recorder?.flush()
        do {
            profile = try await client.get(
                "v1/gamification/profile",
                query: [URLQueryItem(name: "timezone", value: timezone)]
            )
            writeWidgetSnapshot()
        } catch {
            self.error = error.userFacingMessage
        }
        isLoading = false
    }

    /// Widgets show one headline streak/mission, not the full profile — the
    /// longest current streak, and the first not-yet-complete daily mission
    /// (falling back to the first daily mission if all are complete).
    private func writeWidgetSnapshot() {
        guard let widgetStore else { return }
        let topStreak = profile.streaks.max { $0.currentLength < $1.currentLength }.map {
            GamificationWidgetSnapshot.Streak(
                name: $0.name, currentLength: $0.currentLength,
                longestLength: $0.longestLength, graceRemaining: $0.graceRemaining
            )
        }
        let dailyMissions = profile.missions.filter { $0.window == "daily" }
        let dailyChallenge = (dailyMissions.first { $0.progress < $0.target } ?? dailyMissions.first).map {
            GamificationWidgetSnapshot.DailyChallenge(name: $0.name, progress: $0.progress, target: $0.target)
        }
        widgetStore.write(GamificationWidgetSnapshot(topStreak: topStreak, dailyChallenge: dailyChallenge, generatedAt: now()))
        reloadWidgets?()
    }
}

File: android/feature/gamification/src/main/kotlin/com/fatwabot/feature/gamification/GamificationViewModel.kt
package com.fatwabot.feature.gamification

import androidx.lifecycle.ViewModel
import com.fatwabot.core.common.GamificationWidgetSnapshot
import com.fatwabot.core.common.GamificationWidgetSnapshotStore
import com.fatwabot.core.network.AuthenticatedApiClientProtocol
import dagger.hilt.android.lifecycle.HiltViewModel
import java.util.TimeZone
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.serialization.json.Json

/** State machine for the Gamification screen (docs/features/gamification.md).
 * Renders server descriptors verbatim — no client-side scoring/streak math. */
@HiltViewModel
class GamificationViewModel @Inject constructor(
    private val client: AuthenticatedApiClientProtocol,
    private val recorder: GamificationEventRecorder?,
    private val widgetStore: GamificationWidgetSnapshotStore? = null,
    private val onWidgetSnapshotWritten: WidgetRefresh? = null,
) : ViewModel() {
    /** App-supplied hook to trigger Glance updateAll after a new snapshot. */
    fun interface WidgetRefresh {
        fun refresh()
    }

    private val json = Json { ignoreUnknownKeys = true }
    private val timezone: String = TimeZone.getDefault().id

    data class UiState(
        val profile: GamificationProfile = GamificationProfile.EMPTY,
        val isLoading: Boolean = false,
        val error: String? = null,
    )

    private val _state = MutableStateFlow(UiState())
    val state: StateFlow<UiState> = _state.asStateFlow()

    /** Refreshes the profile, keeping the current one visible while the fetch is
     * in flight (the spinner only shows while the profile is still empty). */
    suspend fun load() {
        _state.update { it.copy(isLoading = true, error = null) }
        // Flush any queued events first so a fresh profile reflects them.
        recorder?.flush()
        runCatching {
            val raw = client.getRaw("v1/gamification/profile", mapOf("timezone" to timezone))
            json.decodeFromString(GamificationProfile.serializer(), raw)
        }.fold(
            onSuccess = { profile ->
                _state.update { it.copy(profile = profile, isLoading = false) }
                writeWidgetSnapshot(profile)
            },
            onFailure = { error -> _state.update { it.copy(error = error.toString(), isLoading = false) } },
        )
    }

    /** Widgets show one headline streak/mission, not the full profile — the
     * longest current streak, and the first not-yet-complete daily mission
     * (falling back to the first daily mission if all are complete). */
    private fun writeWidgetSnapshot(profile: GamificationProfile) {
        val store = widgetStore ?: return
        val topStreak = profile.streaks.maxByOrNull { it.currentLength }?.let {
            GamificationWidgetSnapshot.Streak(it.name, it.currentLength, it.longestLength, it.graceRemaining)
        }
        val dailyMissions = profile.missions.filter { it.window == "daily" }
        val dailyChallenge = (dailyMissions.firstOrNull { it.progress < it.target } ?: dailyMissions.firstOrNull())?.let {
            GamificationWidgetSnapshot.DailyChallenge(it.name, it.progress, it.target)
        }
        store.write(GamificationWidgetSnapshot(topStreak, dailyChallenge, System.currentTimeMillis() / 1000))
        onWidgetSnapshotWritten?.refresh()
    }
}
